### ISSUE TITLE
bug: Messages are received out of order from FIFO queue after they became visible again

### ISSUE THREAD

**Issue by Frederick888** [localstack/localstack#4287]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

1. Send messages to a FIFO queue
2. Receive multiple messages (3 in the example)
3. Wait for visibility timeout
4. Receive messages again
5. Messages are received out of order

### Expected Behavior

According to https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/FIFO-queues-understanding-logic.html
> If the consumer detects a failed ReceiveMessage action, it can retry as many times as necessary, using the same receive request attempt ID. Assuming that the consumer receives at least one acknowledgement before the visibility timeout expires, multiple retries don't affect the ordering of messages.

I wasn't using Attempt ID. But based on testing on an actual AWS SQS FIFO queue, the order of messages was still preserved.

### How are you starting LocalStack?

With a docker-compose file

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

```yaml
version: '2.1'

services:
  localstack:
    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
    image: localstack/localstack
    ports:
      - "4566-4599:4566-4599"
      - "${PORT_WEB_UI-8080}:${PORT_WEB_UI-8080}"
    environment:
      - SERVICES=s3,sqs
      - DEBUG=${DEBUG- }
      - DATA_DIR=${DATA_DIR- }
      - PORT_WEB_UI=${PORT_WEB_UI- }
      - LAMBDA_EXECUTOR=${LAMBDA_EXECUTOR- }
      - KINESIS_ERROR_PROBABILITY=${KINESIS_ERROR_PROBABILITY- }
      - DOCKER_HOST=unix:///var/run/docker.sock
      - HOST_TMP_FOLDER=${TMPDIR-/tmp/localstack}
    volumes:
      - "${TMPDIR:-/tmp/localstack}:/tmp/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"
```

    docker-compose up

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)

(save as `create_queue.sh`)
```sh
#!/usr/bin/env bash

[[ -n "$1" ]] && AWS_HOST="$1" || AWS_HOST="localhost"

DEFAULT_SQS_ATTRS='{ "MessageRetentionPeriod": "10800", "ReceiveMessageWaitTimeSeconds": "10", "VisibilityTimeout": "900", "ContentBasedDeduplication": "true" }'
function create_queue() {
    local region="$1"
    local name="$2"
    local dlq="$3"
    local max_recv="$4"
    local vis_timeout="$5"
    [[ -n "$dlq" ]] && dlq="-DLQ"

    local sqs_attributes
    local fifo=""
    if [[ "$name" == *".fifo" ]]; then
        sqs_attributes="$(jq -r -c '.FifoQueue = "true"' <<<"$DEFAULT_SQS_ATTRS")"
        name="${name/.fifo/}"
        fifo=".fifo"
    else
        sqs_attributes="$DEFAULT_SQS_ATTRS"
    fi
    if [[ -n "$vis_timeout" ]]; then
        sqs_attributes="$(jq -r -c --arg vis_timeout "$vis_timeout" '.VisibilityTimeout = $vis_timeout' <<<"$sqs_attributes")"
    fi

    local redrive_policy='{ "maxReceiveCount": "5" }'
    if [[ -n "$max_recv" ]]; then
        redrive_policy="$(jq -r -c --arg max_recv "$max_recv" '.maxReceiveCount = $max_recv' <<<"$redrive_policy")"
    fi
    if [[ -n "$dlq" ]]; then
        aws --endpoint-url="http://$AWS_HOST:4566" --region "$region" sqs create-queue --queue-name "$name$dlq$fifo" --attributes "$sqs_attributes"
        redrive_policy="$(jq -r -c "$(printf '.deadLetterTargetArn = "arn:aws:sqs:%s:000000000000:%s-DLQ%s"' "$region" "$name" "$fifo")" <<<"$redrive_policy")"
        sqs_attributes="$(jq -r -c --arg policy "$redrive_policy" '.RedrivePolicy = $policy' <<<"$sqs_attributes")"
    fi
    aws --endpoint-url="http://$AWS_HOST:4566" --region "$region" sqs create-queue --queue-name "$name$fifo" --attributes "$sqs_attributes"
}
```

```sh
$ . ./create_queue.sh
$ create_queue us-east-1 test1.fifo 1 50 60
{
    "QueueUrl": "http://$AWS_HOST:4566/000000000000/test1-DLQ.fifo"
}
{
    "QueueUrl": "http://$AWS_HOST:4566/000000000000/test1.fifo"
}
$ aws --endpoint-url=http://$AWS_HOST:4566 sqs get-queue-attributes --queue-url http://$AWS_HOST:4566/000000000000/test1.fifo
{
    "Attributes": {
        "ApproximateNumberOfMessages": "0",
        "ApproximateNumberOfMessagesDelayed": "0",
        "ApproximateNumberOfMessagesNotVisible": "0",
        "CreatedTimestamp": "1626137940.978319",
        "DelaySeconds": "0",
        "LastModifiedTimestamp": "1626137940.978319",
        "MaximumMessageSize": "262144",
        "MessageRetentionPeriod": "10800",
        "QueueArn": "arn:aws:sqs:us-east-1:000000000000:test1.fifo",
        "RedrivePolicy": "{\"maxReceiveCount\":50,\"deadLetterTargetArn\":\"arn:aws:sqs:us-east-1:000000000000:test1-DLQ.fifo\"}",
        "ReceiveMessageWaitTimeSeconds": "10",
        "VisibilityTimeout": "60",
        "FifoQueue": "true",
        "ContentBasedDeduplication": "true"
    }
}
$ aws --endpoint-url=http://$AWS_HOST:4566 sqs send-message --queue-url http://$AWS_HOST:4566/000000000000/test1.fifo --message-body '{"foo": "bar4"}' --message-group-id 'helloworld'
{
    "MD5OfMessageBody": "e600d75d2e7c74c5c2299fab1762bc50",
    "MessageId": "0fb8fe29-6b73-aa21-f845-03bdac37150a"
}
$ aws --endpoint-url=http://$AWS_HOST:4566 sqs send-message --queue-url http://$AWS_HOST:4566/000000000000/test1.fifo --message-body '{"foo": "bar5"}' --message-group-id 'helloworld'
{
    "MD5OfMessageBody": "bbc312e807f097092fbaa4ea30859c5e",
    "MessageId": "5a529b21-53e9-0154-05df-cfafd5c96e6e"
}
$ aws --endpoint-url=http://$AWS_HOST:4566 sqs send-message --queue-url http://$AWS_HOST:4566/000000000000/test1.fifo --message-body '{"foo": "bar6"}' --message-group-id 'helloworld'
{
    "MD5OfMessageBody": "21b93ade231a56d0babcd3f4100a060e",
    "MessageId": "9dc8d2c8-ccf7-ae90-bc40-1039fac2376c"
}
$ aws --endpoint-url=http://$AWS_HOST:4566 sqs receive-message --queue-url http://$AWS_HOST:4566/000000000000/test1.fifo --attribute-names All --message-attribute-names All --max-number-of-messages 10
{
    "Messages": [
        {
            "MessageId": "0fb8fe29-6b73-aa21-f845-03bdac37150a",
            "ReceiptHandle": "kfzfupygrguevwxpmoqewrajxtrpuedcptmmxsnqbchlczneuduxpwtnntibfqzgnkbbqwjjukghgfgjloicgpylulswbrtihdhmfzracmfsvuygczwaakuxfoeecduzoaxbslnxdbjgrivpblnzmienvtuycsvfxawnffacxmbpgssteoopsslmx",
            "MD5OfBody": "e600d75d2e7c74c5c2299fab1762bc50",
            "Body": "{\"foo\": \"bar4\"}",
            "Attributes": {
                "SenderId": "AIDAIT2UOQQY3AUEKVGXU",
                "SentTimestamp": "1626138003146",
                "ApproximateReceiveCount": "1",
                "ApproximateFirstReceiveTimestamp": "1626138037904",
                "MessageDeduplicationId": "e03bde47f4d505a2b7d6cc082e0424fed2fad73c67f0f16d0602db4d49dd0fc4",
                "MessageGroupId": "helloworld"
            }
        },
        {
            "MessageId": "5a529b21-53e9-0154-05df-cfafd5c96e6e",
            "ReceiptHandle": "cbecbinornqpeegsesrpukgnitwnuzspyxrfgzzauzwcgsqsgdcsicfjsepnqxggokhtqildwmjltsuvgjtbkvtkgnqvojgsqzljlkjabyeobwbcanftyqfopuxqnqebudxunjnsnkfydifhfdjbijteenqjyylgzdercbcvsknxrkalxkeksvcaa",
            "MD5OfBody": "bbc312e807f097092fbaa4ea30859c5e",
            "Body": "{\"foo\": \"bar5\"}",
            "Attributes": {
                "SenderId": "AIDAIT2UOQQY3AUEKVGXU",
                "SentTimestamp": "1626138007641",
                "ApproximateReceiveCount": "1",
                "ApproximateFirstReceiveTimestamp": "1626138037905",
                "MessageDeduplicationId": "ee6e00124d6352962c5ad9bd21738086b814eec5b739da18c3aaa7ebe66d9aea",
                "MessageGroupId": "helloworld"
            }
        },
        {
            "MessageId": "9dc8d2c8-ccf7-ae90-bc40-1039fac2376c",
            "ReceiptHandle": "rgfhtlxitiwmhuqqbwmxlossabddczvqjkwaznvknusxqrokfndqokcewkkmztmimgkpyvvmqynnpmjvjyksoymnohxitznkwnrpfxlusjtsijbvncbqwtidxfweorzvnqkjwyzurkjqnxcmsojfqwdvoquckhwtexmuoplqetvksssxtniwhfrhx",
            "MD5OfBody": "21b93ade231a56d0babcd3f4100a060e",
            "Body": "{\"foo\": \"bar6\"}",
            "Attributes": {
                "SenderId": "AIDAIT2UOQQY3AUEKVGXU",
                "SentTimestamp": "1626138011355",
                "ApproximateReceiveCount": "1",
                "ApproximateFirstReceiveTimestamp": "1626138037906",
                "MessageDeduplicationId": "606514bb7bae1a6bdde8176c07d192b8d7d596c8b33cd611afaf5fc53f2c00f4",
                "MessageGroupId": "helloworld"
            }
        }
    ]
}
$ aws --endpoint-url=http://$AWS_HOST:4566 sqs receive-message --queue-url http://$AWS_HOST:4566/000000000000/test1.fifo --attribute-names All --message-attribute-names All --max-number-of-messages 10
{
    "Messages": [
        {
            "MessageId": "9dc8d2c8-ccf7-ae90-bc40-1039fac2376c",
            "ReceiptHandle": "jdbonffjtigvdhabkbyhebzxthmoyqpjcicylxsufullmsjkywvrkwejyvpddaswmvznkggrbflcaoqhayhokflzfastgvwltmyjuhwbpaslbvguagnojvmmvqoxwoehwtpfwesjkpbotpqawggysosqwtpkircawnehpzxustijeyhpezrutifwa",
            "MD5OfBody": "21b93ade231a56d0babcd3f4100a060e",
            "Body": "{\"foo\": \"bar6\"}",
            "Attributes": {
                "SenderId": "AIDAIT2UOQQY3AUEKVGXU",
                "SentTimestamp": "1626138011355",
                "ApproximateReceiveCount": "2",
                "ApproximateFirstReceiveTimestamp": "1626138037906",
                "MessageDeduplicationId": "606514bb7bae1a6bdde8176c07d192b8d7d596c8b33cd611afaf5fc53f2c00f4",
                "MessageGroupId": "helloworld"
            }
        }
    ]
}
```

### Environment

```markdown
- OS: Arch Linux
- LocalStack version: 0.12.15
- LocalStack build date: 2021-07-12
- LocalStack build git hash: 01e81ec3
```

**Comment by Frederick888** [comment]
It seems ElasticMQ's behaviour is correct but I'm having trouble using it with AWS Java SDK:

```
Caused by: com.amazonaws.SdkClientException: Unable to unmarshall response (Duplicated keys (ContentBasedDeduplication) are provided.). Response Code: 200, Response Text:
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleResponse(AmazonHttpClient.java:1758)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleSuccessResponse(AmazonHttpClient.java:1454)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1369)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1145)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:802)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:770)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:744)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:704)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:686)
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:550)
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:530)
        at com.amazonaws.services.sqs.AmazonSQSClient.doInvoke(AmazonSQSClient.java:2271)
        at com.amazonaws.services.sqs.AmazonSQSClient.invoke(AmazonSQSClient.java:2238)
        at com.amazonaws.services.sqs.AmazonSQSClient.invoke(AmazonSQSClient.java:2227)
        at com.amazonaws.services.sqs.AmazonSQSClient.executeGetQueueAttributes(AmazonSQSClient.java:1154)
        at com.amazonaws.services.sqs.AmazonSQSClient.getQueueAttributes(AmazonSQSClient.java:1122)
        at com.amazonaws.services.sqs.AmazonSQSClient.getQueueAttributes(AmazonSQSClient.java:1166)
        at au.com.bglcorp.digitalsigning.service.impl.AmazonSqsServiceImpl.getQueueAttributes(AmazonSqsServiceImpl.java:137)
        at au.com.bglcorp.digitalsigning.service.FifoQueueMessageManagementService.validateQueueAttributes(FifoQueueMessageManagementService.java:51)
        at au.com.bglcorp.digitalsigning.service.impl.NotificationQueueMessageManagementService.postConstruct(NotificationQueueMessageManagementService.java:45)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:366)
        at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:311)
        at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:134)
        ... 88 common frames omitted
Caused by: java.lang.IllegalArgumentException: Duplicated keys (ContentBasedDeduplication) are provided.
        at com.amazonaws.services.sqs.model.GetQueueAttributesResult.addAttributesEntry(GetQueueAttributesResult.java:91)
        at com.amazonaws.services.sqs.model.transform.GetQueueAttributesResultStaxUnmarshaller.unmarshall(GetQueueAttributesResultStaxUnmarshaller.java:92)
        at com.amazonaws.services.sqs.model.transform.GetQueueAttributesResultStaxUnmarshaller.unmarshall(GetQueueAttributesResultStaxUnmarshaller.java:32)
        at com.amazonaws.http.StaxResponseHandler.handle(StaxResponseHandler.java:117)
        at com.amazonaws.http.StaxResponseHandler.handle(StaxResponseHandler.java:42)
        at com.amazonaws.http.response.AwsResponseHandlerAdapter.handle(AwsResponseHandlerAdapter.java:69)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleResponse(AmazonHttpClient.java:1734)
        ... 114 common frames omitted
```

**Comment by thrau** [comment]
The issue should be fixed with our new provider deployment. Could you please pull the latest image and try again?

**Comment by thrau** [comment]
Thanks for the detailed report @Frederick888, will try and reproduce and get back to you

**Comment by thrau** [comment]
The latest issue was addressed in #6313

**Comment by thrau** [comment]
For the original issue we wrote a test a while back that should now be green #6578
